### PR TITLE
make timestamp string format cassandra compatible

### LINKED ISSUE #7997: toJson() is missing a timezone on timestamp
When running the `toJson()` function on a timestamp column, the result in Cassandra looks like `"2014-01-01 00:00:00.000Z"` while in Scylla it is `"2014-01-01T00:00:00"`. Scylla's output has two problems:

1. The "T" instead of space is indeed one of the allowed _input_ formats for timestamp (see https://cassandra.apache.org/doc/latest/cql/types.html#timestamps) but it was not the format chosen by Cassandra for `toJson()`'s _output_.

2. When a timezone specification is missing, the aforementioned Cassandra documentation states that "_the date will be interpreted as being in the time zone under which the coordinating Cassandra node is configured_", and that is not useful so "_it is recommended that the time zone always be specified for timestamps when feasible._". Cassandra's `toJson()` was missing the timezone in the past - see https://issues.apache.org/jira/browse/CASSANDRA-11137 - but it was fixed in Cassandra commit 88f22b9692c6fdddf837556f13140d949afe0d28 from 2016 - which landed in Cassandra 2.2.7, 3.0.6, 3.6 and of course 4.0. 

This issue is reproduced by `cassandra_tests/validation/entities/json_test.py`::`testToJsonFct`, a translated Cassandra test which among many other things tests `toJson()` of a timestamp column.

### LINKED ISSUE #14518: CAST from timestamp to string not same as Cassandra on zero milliseconds
This issue was previously noted in the discussion on pull request https://github.com/scylladb/scylladb/issues/13588 and in issue #7997, but we didn't have a separate issue about this until now.

When x is a `timestamp` column, and we use `SELECT CAST(x AS text)`, the format of the output looks like ` 2017-12-27T11:57:42.500Z`

It turns out that if the millseconds part is zero (the timestamp is a whole second), Scylla drops the ".000Z" entirely - i.e., drops both the zero milliseconds **and** also the timezone part ("Z"). This makes little sense (why drop the timezone only on whole second) and incompatible with Cassandra.

The following translated Cassandra unit test reproduces this issue:
* `cassandra_tests/functions/cast_fcts_test.py`::`testTimeCastsInSelectionClause`

### PR DESCRIPTION
**Pull request by alezzqz**
when we convert timestamp into string it must look like: '2017-12-27T11:57:42.500Z'                
it concerns any conversion except JSON timestamp format                                                                
JSON string has space as time separator and must look like: '2017-12-27 11:57:42.500Z'
both formats always contain milliseconds and timezone specification

Fixes #14518
Fixes #7997

### CODE REVIEW

File: types/types.cc
Comment (tchaikov) [inline types/types.cc:100]: this changes the output like: 
```diff
- "2011-02-03T04:05:00"
+ "2011-02-03T04:05:00.000Z"
```

so if we want to go in this direction. we might want to update the tests in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 accordingly

File: types/types.cc
Comment (tchaikov) [inline types/types.cc:99]: what is this for?

File: types/types.cc
Comment (alezzqz) [reply types/types.cc:100]: sorry, 404

File: types/types.cc
Comment (alezzqz) [reply types/types.cc:99]: to round to 3 decimal places

File: types/types.cc
Comment (denesb) [reply types/types.cc:100]: @tchaikov scylla-dtest.git is private repository, it is not available to external contributors (not sure why).

Comment (tchaikov) [reply types/types.cc:100]: sorry, i just realized this.

File: types/types.cc
Comment (tchaikov) [reply types/types.cc:99]: `millis` is a `long int`, so if we multiply it by 1000 if it does not overflow, the result would be `millis * 1000`. after dividing it by `1000`, the result should be `millis`. am i missing anything? or could you please raise an example?

Comment (alezzqz) [reply types/types.cc:99]: oops, my mistake, will fix
thanks

**Comment by alezzqz** [comment]
changes:
- removed millis rounding cause it has no sense: the value can't have more than 3 digits
- refactored `time_point_to_string` function to simplify it

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2551/

The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.

**Comment by alezzqz** [comment]
> The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.

I'd fix these tests but have no access to them (

**Comment by denesb** [comment]
> > The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.
> 
> I'd fix these tests but have no access to them (

I will look into fixing these tests today. Will keep you posted.

**Comment by denesb** [comment]
dtest PR is here: https://github.com/scylladb/scylla-dtest/pull/3316, once it goes in we can merge this PR (provided no new failures pop up).

**Comment by denesb** [comment]
I will re-kick the CI once https://github.com/scylladb/scylla-dtest/pull/3316 goes in.

**Comment by nyh** [comment]
@alezzqz if I understood correctly from @denesb, this patch also fixes the behavor of JSON. So does it also "Fixes #7997"? If it does, please add this phrase to the commit message, and also un-xfail all the tests we have for 7997 (if they, hopefully, start passing).

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good, but please check if it also Fixes #7997 - I believe it should (and if it doesn't, maybe this patch has a bug). If it makes the xfail-ing tests for #7997 pass, this patch should also remove those xfail tags. Thanks.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2585/

Looks like my test fixes didn't propagate yet. @fruch how much time does it take for https://github.com/scylladb/scylla-dtest/pull/3316 to make its way into CI?

**Comment by fruch** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2585/
> 
> Looks like my test fixes didn't propagate yet. @fruch how much time does it take for https://github.com/scylladb/scylla-dtest/pull/3316 to make its way into CI?

It needs to pass next as any other scylla commit

**Comment by alezzqz** [comment]
> @alezzqz if I understood correctly from @denesb, this patch also fixes the behavor of JSON. So does it also "Fixes #7997"? If it does, please add this phrase to the commit message, and also un-xfail all the tests we have for 7997 (if they, hopefully, start passing).

Yes, it seems JSON uses the same function to convert to string so it's behavior should be fixed as well.
I will mention it in commit description.

File: test/boost/expr_test.cc
Comment (alezzqz) [reply test/boost/expr_test.cc:166]: In this test we compare timestamp value converted to string so it's just a way of CAST but you're right, we need to make commit message more correct. I will try to do.
Concerning other tests using +0000 - as I see all of them use this string values to parse and create timestamps and that format is correct for cassandra too. Do you think we'd better replace them to .000Z? It makes sense...

File: types/types.cc
Comment (nyh) [reply types/types.cc]: > @tchaikov scylla-dtest.git is private repository, it is not available to external contributors (not sure why).

It was like that from day one, I also don't know why. 
@alezzqz you don't need to worry too much, the better (in my opinion) functional test frameworks are all in scylla.git itself and all open-source.

File: test/boost/expr_test.cc
Comment (alezzqz) [reply test/boost/expr_test.cc:166]: Agree, will change tests now

Comment (nyh) [reply test/boost/expr_test.cc:166]: I'm not sure what you mean - I said that I thought the changes you already did to the tests were fine, you don't need to change them again. I was surprised why the **old** result (before your patch) was the way it was, but the new result seems fine.

Comment (alezzqz) [reply test/boost/expr_test.cc:166]: Okay, seems I didn't understand you properly. So I will change commit description only

Comment (nyh) [reply test/boost/expr_test.cc:166]: And please (unrelated to this thread of discussion) check that the xfailing tests for #7997 now pass, and remove their xfail mark.

Comment (alezzqz) [reply test/boost/expr_test.cc:166]: the problem is the test reproduces few issues (#7972, #7988, #7997, #8001) but we fixed only #7997 so removing xfail mark will bring to failing test. Maybe we should extract part for #7997 into separate test?

Comment (nyh) [reply test/boost/expr_test.cc:166]: You're right. So you can't remove the xfail, sorry. Just please remove the word "#7997" from the xfail line, and that's it (leave the "Reproduces:" line).
I would indeed be nice if you could extract the few lines that test 7997 to a separate test, e.g., put it in cql-pytest/test_json.py. Then you can be sure your patch fixes 7997 (and then please add "Fixes #7997" to the commit message). Thanks.

**Comment by alezzqz** [comment]
changes v3:
- changed commit and PR description to be more correct
- extracted reproducing #7997 code from unit test json_test.py:testToJsonFct to new test json_test.py::testTimestampToJson

**Comment by denesb** [comment]
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

Kicked CI.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2610/

The same json tests are still failing. Futhermore, a scylla sstable dump test is failing. I will look into it, as I will be off tomorrow, this will probably only happen on Monday.

**Comment by nyh** [comment]
Please split the Fixes line into two separate Fixes lines, or github won't recognize these magic keywords. As the github documentation https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue explains: "_Use full syntax for each issue_".

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (nyh) [inline test/cql-pytest/cassandra_tests/validation/entities/json_test.py:856]: Please do not modify tests in the cassandra_tests/ directory - these are tests translated from the Cassandra unit tests, and I don't want to add new tests or delete pieces of tests in these files. Please leave this test unchanged, and the new test you added, testTimestampToJson, please don't put it in the cassandra_tests/ directory but rather in `cql-pytest/test_json.py`. The comment in this file explains its purpose, and among other things it includes : _small tests aiming to reproduce bugs discovered by bigger Cassandra tests._

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (nyh) [inline test/cql-pytest/cassandra_tests/validation/entities/json_test.py:928]: Please move this test to `cql-pytest/test_json.py` (for reasons I explained above).
Please note that that the coding convention in the general cql-pytest/ directory is different from cassandra_tests - in cassandra_tests I used a bunch of convenience functions to help me port the Java unit tests, but these aren't available and aren't needed in cql-pytest/test_json.py.

But beyond that, you should look at the content of the test and simplify it the bare minimum... This test does NOT need the create_type (I don't know why you needed it), and it shouldn't even create a new table just for this test - it should use the "table1" table that all the tests in the same file are using. This reduces the runtime of such a test from something like 0.1 seconds to something like 0.001 seconds, and makes me happy ;-)

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (nyh) [inline test/cql-pytest/cassandra_tests/validation/entities/json_test.py:929]: This type isn't used in this test and can be removed.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good, but I still have some requests (sorry...) about the tests.
We also need to make the dtests pass before we can commit this PR - and @denesb said he'll do it.

I guess that like the C++ boost tests, the dtests "enshrined" a bunch of wrong (or at least, non-Cassandra-compatible) outputs, and nobody ever tested these tests against Cassandra so we didn't know. This is why I always insist that test/cql-pytest tests should also be run against Cassandra.

Also, as I asked (and explained) in previous comment, please split the Fixes line in the commit message into two separate lines. Thanks.

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (alezzqz) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py:935]: Thanks. I run cassandra in docker container, is it possible to run this script using cassandra in container?

Comment (nyh) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py:935]: The run-cassandra script is not much more than a convenience script that runs Cassandra for you in a temporary directory, and then runs "pytest" to run all the tests. If you know how to run Cassandra yourself (via docker or however you're used to running Cassandra), then just run it yourself, and then do something like
```
cd cqlpytest
pytest json_test.py::some_function_name
```

If your Cassandra isn't listening on localhost:9042 but rather on some other IP address or port (possibly even on another machine), just pass the "--host" and/or "--port" option to pytest, e.g.,
```
pytest --host 1.2.3.4 --port 4567 json_test.py::some_function_name
```

It's very easy and convenient.

Comment (alezzqz) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py]: thanks, will try

**Comment by alezzqz** [comment]
changes v4:
- returned back all changes to imported cassandra tests
- created new test for #7997 in test/cql-pytest/test_json.py
- changed "fixed" lines in commit descriptions
- rebased on latest master

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (alezzqz) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py]: I've got an idea how to fix problem with JSON string format but not sure we should do it in this patch...

Comment (alezzqz) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py]: I've got an idea how to fix JSON format, will add to this patch if you don't want to split these fixes into different patches

File: types/types.cc
Comment (tchaikov) [inline types/types.cc:91]: note to myself, we can drop the workaround in https://github.com/scylladb/scylla-dtest/blob/0324001e1d5275eb6d6a120018d51e7e0d4813e0/scylla_sstable_test.py#L176, once this change lands.

Comment (denesb) [reply types/types.cc:91]: In fact, the above test fails with this PR currently, so we will have to modify said test to work both with and without this PR, before this can go it. A CI is running currently and we should soon know the up-to-date list of dtest that sill require patching.

It is such a pain to do changes that require dtest changes too...

Comment (tchaikov) [reply types/types.cc:91]: yeah, i see. in the small time window. we need to at least appease both versions. lemme at least create a patch for the scylla_sstable_test.py . i really want to replace sstableloader in favor of `scylla dump-data`. as, IMO, a partial uuid-generation solution is not a solution ready for GA.

this should help us to partially address https://github.com/scylladb/scylla-tools-java/issues/333 -- one tiny step to ditch `tools/java`. lemme know if i can help.

Comment (tchaikov) [reply types/types.cc:91]: the dtest finished, see https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/

- [ ] [scylla_sstable_test.TestScyllaSstableDumpataAllDatatypes.test_sstabledump_all_datatypes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/testReport/junit/scylla_sstable_test/TestScyllaSstableDumpataAllDatatypes/FullDtest___full_split000___test_sstabledump_all_datatypes/) should be addressed by https://github.com/scylladb/scylla-dtest/pull/3328
- [ ] [update_cluster_layout_tests.TestUpdateClusterLayout.test_add_node_with_large_partition4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split001___test_add_node_with_large_partition4/): i am still on it

**Comment by alezzqz** [comment]
changes v5:
- added function to support cassandra compatible format for JSON timestamp strings
- modified commit and PR description

File: test/cql-pytest/cassandra_tests/validation/entities/json_test.py
Comment (alezzqz) [reply test/cql-pytest/cassandra_tests/validation/entities/json_test.py]: I've added a function to support JSON timestamp string format. It's the simplest way I found.
I tried to add t`o_json_string() `method to `timestamp_date_base_class` and maybe it would look better a little but it required a lot of linked changes and I've chosen another way.
All tests are passed locally, including our json tests on Cassandra as we discussed above

**Review by nyh — APPROVED** [review]

File: cql3/type_json.cc
Comment (nyh) [reply cql3/type_json.cc:492]: But neither of these changes are critical. Anyway, now I'm waiting for CI to pass (I started it now).

**Comment by denesb** [comment]
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

Kicked CI.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/

[Tests / Sanity Tests / json_test.TestsToJsonSelect.test_basic_data_types](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/testReport/junit/json_test/TestsToJsonSelect/Tests___Sanity_Tests___test_basic_data_types/)
    [Tests / Sanity Tests / json_test.TestsFromJsonInsert.test_basic_data_types](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/testReport/junit/json_test/TestsFromJsonInsert/Tests___Sanity_Tests___test_basic_data_types/)

I don't understand why these failed, I already prepared these tests. I will look into this.

**Comment by tchaikov** [comment]
@denesb 

```diff
- "2011-02-03T04:05:00.000Z"
+ "2011-02-03 04:05:00.000Z"
```

File: types/types.cc
Comment (tchaikov) [inline types/types.cc:90]: why quote `use_time_separator` with parenthesis ? also, a single char would suffice, no need to use a string literal here.

**Comment by denesb** [comment]
> @denesb
> 
> ```diff
> - "2011-02-03T04:05:00.000Z"
> + "2011-02-03 04:05:00.000Z"
> ```

Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.

File: cql3/type_json.cc
Comment (alezzqz) [reply cql3/type_json.cc:492]: > Looks good, just one thought: If the function is called timestamp_to_json_string, maybe it should return the full JSON string, i.e., with the double quotes? It can either call quote_json_string itself like you did, or just add the two quotes in the beginning and the end (there is no need to escape anything because you know you didn't use any special characters).

I thought about that but decided to keep code's common style: when we wrapping strings in quotes in visitor class.
I could do it just inside function if you think it's better.

> Another nitpick: Since this new function timestamp_to_json_string() is used only in this source file, maybe its definition can be in this file and not a different file (and it can be a static function)? Or it needs non-public APIs to be implemented?

Oh, that's would be much better, I agree. And I wanted to do that way. But this function uses things from different module so we should export something anyway. Better of two evils...
I could try implement dependent functionality just in place if you think it would be better

**Comment by tchaikov** [comment]
> > @denesb
> > ```diff
> > - "2011-02-03T04:05:00.000Z"
> > + "2011-02-03 04:05:00.000Z"
> > ```
> 
> Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.

this is exactly what @nyh complained at https://github.com/scylladb/scylladb/pull/14726#discussion_r1273773283 . Cassandra use a non-ISO 8601 format when serializing a datetime to JSON, but it format it with ISO 8601 when it prints it as non-JSON. see https://github.com/apache/cassandra/blob/d08fd1ebd83edc4730316c9630e60b8c0b0be89d/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L108-L126

**Comment by denesb** [comment]
> > > @denesb
> > > ```diff
> > > - "2011-02-03T04:05:00.000Z"
> > > + "2011-02-03 04:05:00.000Z"
> > > ```
> > 
> > 
> > Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.
> 
> this is exactly what @nyh complained at [#14726 (comment)](https://github.com/scylladb/scylladb/pull/14726#discussion_r1273773283) . Cassandra uses a non-ISO 8601 format when serializing a datetime to JSON, but it formats it with ISO 8601 when it prints it as non-JSON. see https://github.com/apache/cassandra/blob/d08fd1ebd83edc4730316c9630e60b8c0b0be89d/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L108-L126

:(
Ok, let me prepare a dtest PR.

**Comment by denesb** [comment]
https://github.com/scylladb/scylla-dtest/pull/3341

File: types/types.cc
Comment (denesb) [reply types/types.cc:90]: Agree, no need to send a new version for this.

Comment (tchaikov) [reply types/types.cc:90]: yeah. please consider it a nitpick. i will create a follow-up change once this change lands.

**Comment by denesb** [comment]
All the prerequisites got in, CI passed, queuing.

**Comment by nyh** [comment]
I want to comment that in retrospect, this fix is more important than we had realized and fixes a third bug which we didn't have listed - and had we known about it it might have caused us to backport this fix.

The third bug is that Scylla couldn't read the timestamps that it itself wrote. It wrote timestamps like `2009-12-17T00:26:29.805000` which then it couldn't read because 6 digits of precision aren't allowed for timestamps (Scylla reports "Milliseconds Exceed Expected Length"). We had a customer complaining about exactly this. This is more serious than the minor Cassandra compatibility issues of https://github.com/scylladb/scylladb/issues/14518
or https://github.com/scylladb/scylladb/issues/7997.

You can see in this patch all the "805000" being changed to "805" in the tests, but the importance of this was not realized, and it wasn't even mentioned in the commit message or any of the issues it referred to.

CC @ptrsmrn @alezzqz

**Comment by nyh** [comment]
Opened https://github.com/scylladb/scylladb/issues/16575 to consider further backports

**Pull request closed without merging**
